### PR TITLE
[FLINK-33767] Implement restore tests for TemporalJoin node

### PR DESCRIPTION
**Pull request by jnh5y**
## What is the purpose of the change

Implement restore tests for TemporalJoin node

## Verifying this change

This change added tests and can be verified as follows:

* Added restore tests for TemporalJoin node which verifies the generated compiled plan with the saved compiled plan

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/test/program/TemporalFunctionTestStep.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public final FunctionPersistence persistence;
+    public final FunctionBehavior behavior;
+    public final String name;
+    public final String table;
+    public final Expression timeAttribute;
+    public final Expression primaryKey;
+
+    TemporalFunctionTestStep(
```
Comment (jnh5y) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/test/program/TemporalFunctionTestStep.java:47]: @twalthr @dawidwys Any ideas on unifying temporal functions with regular functions? 

I'm open for feedback on how to improve this class and how it integrates with the tests.

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/test/program/TemporalFunctionTestStep.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    public void apply(TableEnvironment env) {
+        TemporalTableFunction function =
+                env.from(table).createTemporalTableFunction(timeAttribute, primaryKey);
+        if (behavior == FunctionBehavior.SYSTEM) {
+            if (persistence == FunctionPersistence.TEMPORARY) {
+                env.createTemporarySystemFunction(name, function);
+
+            } else {
+                throw new UnsupportedOperationException("System functions must be temporary.");
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/test/program/TemporalFunctionTestStep.java:75]: I wouldn't add options which are not possible. Temporal functions can only ever be temporary.  Other than that I'd say it's fine.

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/test/program/TemporalFunctionTestStep.java (hunk 1 above)
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/test/program/TemporalFunctionTestStep.java:47]: I think this is legacy way of performing a temporal join, so I believe it's fine to have a separate step.

**Review by dawidwys — COMMENTED** [review]
Other than the comments on the temporal function test step looks good.

**Review by dawidwys — APPROVED** [review]
lgtm

**Pull request closed without merging**